internal/lookoutingesterv2/lookoutdb/insertion.go

```diff
@@ -0,0 +1,884 @@
+
+	// Jobs need to be ingested first as other updates may reference these
+	l.CreateJobs(ctx, instructions.JobsToCreate)
+
+	// Now we can job updates, annotations and new job runs
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+	go func() {
+		defer wg.Done()
+		l.UpdateJobs(ctx, jobsToUpdate)
+	}()
+	go func() {
+		defer wg.Done()
+		l.CreateJobRuns(ctx, instructions.JobRunsToCreate)
+	}()
+	go func() {
+		defer wg.Done()
+		l.CreateUserAnnotations(ctx, instructions.UserAnnotationsToCreate)
+	}()
+
+	wg.Wait()
+
+	// Finally, we can update the job runs
+	l.UpdateJobRuns(ctx, jobRunsToUpdate)
+	return nil
+}
+
+func (l *LookoutDb) CreateJobs(ctx context.Context, instructions []*model.CreateJobInstruction) {
+	if len(instructions) == 0 {
+		return
+	}
+	err := l.CreateJobsBatch(ctx, instructions)
+	if err != nil {
+		log.Warnf("Creating jobs via batch failed, will attempt to insert serially (this might be slow).  Error was %+v", err)
+		l.CreateJobsScalar(ctx, instructions)
+	}
+}
+
+func (l *LookoutDb) UpdateJobs(ctx context.Context, instructions []*model.UpdateJobInstruction) {
+	if len(instructions) == 0 {
+		return
+	}
+	instructions = l.filterEventsForCancelledJobs(ctx, l.db, instructions, l.metrics)
+	err := l.UpdateJobsBatch(ctx, instructions)
+	if err != nil {
+		log.Warnf("Updating jobs via batch failed, will attempt to insert serially (this might be slow).  Error was %+v", err)
+		l.UpdateJobsScalar(ctx, instructions)
+	}
+}
+
+func (l *LookoutDb) CreateJobRuns(ctx context.Context, instructions []*model.CreateJobRunInstruction) {
+	if len(instructions) == 0 {
+		return
+	}
+	err := l.CreateJobRunsBatch(ctx, instructions)
+	if err != nil {
+		log.Warnf("Creating job runs via batch failed, will attempt to insert serially (this might be slow).  Error was %+v", err)
+		l.CreateJobRunsScalar(ctx, instructions)
+	}
+}
+
+func (l *LookoutDb) UpdateJobRuns(ctx context.Context, instructions []*model.UpdateJobRunInstruction) {
+	if len(instructions) == 0 {
+		return
+	}
+	err := l.UpdateJobRunsBatch(ctx, instructions)
+	if err != nil {
+		log.Warnf("Updating job runs via batch failed, will attempt to insert serially (this might be slow).  Error was %+v", err)
+		l.UpdateJobRunsScalar(ctx, instructions)
+	}
+}
+
+func (l *LookoutDb) CreateUserAnnotations(ctx context.Context, instructions []*model.CreateUserAnnotationInstruction) {
+	if len(instructions) == 0 {
+		return
+	}
+	err := l.CreateUserAnnotationsBatch(ctx, instructions)
+	if err != nil {
+		log.Warnf("Creating user annotations via batch failed, will attempt to insert serially (this might be slow).  Error was %+v", err)
+		l.CreateUserAnnotationsScalar(ctx, instructions)
+	}
+}
+
+func (l *LookoutDb) CreateJobsBatch(ctx context.Context, instructions []*model.CreateJobInstruction) error {
+	return l.withDatabaseRetryInsert(func() error {
+		tmpTable := uniqueTableName("job")
+
+		createTmp := func(tx pgx.Tx) error {
+			_, err := tx.Exec(ctx, fmt.Sprintf(`
+				CREATE TEMPORARY TABLE %s 
```

Comment:
somethinn weird is going on with the formatting here!

Reply:
and with all the other sql queries!

Reply:
Very weird - looks correct in GoLand